### ISSUE TITLE
Static destructors are called before atexit() in c++

### ISSUE THREAD

**Issue by Samuel-DEVULDER** [bebbo/libnix#29]
Consider that piece of C++ code
```
#include <iostream>
#include <stdlib.h>

using namespace std;

class Cls  {
	public:
	Cls() {
		printf("Created\n");
	}
	~Cls() {
		printf("Destroyed\n");
	}
} cls = Cls();

void onExit(void)
{
	printf("onExit\n");
}

int main() 
{
	atexit(onExit);
    cout << "Hello, World!" << endl;
    return 0;
}

```
And compile & run it on pc with:
```
$ g++ test.cpp -o a.out && ./a.out
Created
Hello, World!
onExit
Destroyed

```
Now compile it  with amiga-gcc's g++ compiler:
`$ /opt/amiga/bin/m68k-amigaos-g++ test.cpp -o a.out -noixemul
`
And run it on the amiga. You'll get this:
```
1> a.out
Created
Hello, World!
Destroyed
onExit

```
As you can see the onExit() function is called after the static instance has been destroyed, contrary to the PC version. This mean that in real applications, some atexit() cleanup code might refer to destroyed object on the amiga resulting into random crashes (I observe this with devilutionx on my github for instance).

According to [this discussion](https://stackoverflow.com/questions/16010083/order-between-destruction-of-global-object-and-atexit-in-c), the PC behavior is correct whereas the amiga behavior is broken.

As a side note: when compiling with ixemul, we don't see the call to onExit(), maybe because stdout has been closed. So we can't tell if this bug is related to libnix only or spans over ixemul as well.

**Comment by bebbo** [comment]
please test
```
 > vamos x
Created
Hello, World!
onExit
Destroyed
```

**Comment by Samuel-DEVULDER** [comment]
I've just recompiled amiga-gcc, and run the test. It seem to work fine. Great! #

**Comment by Samuel-DEVULDER** [comment]
Hmm maybe I spoke too quickly. The test-case works fine, but now devilutionx crashes on startup :(

**Comment by bebbo** [comment]
so something is not initialized at that earlier time?

**Comment by Samuel-DEVULDER** [comment]
It is hard to tell. Simple programs works, but devilutionx crashes. I have tried to revert to prio=100 as before in ___initcpp.c, but devilution still crashes. It might be another change in gcc as well (My version was compiled in ~20th of last october).

**Comment by Samuel-DEVULDER** [comment]
Okay.. the issue is is now found. Devilutionx uses libSDL_mixer.a that was compiled on oct 31th and for some reason, that old version crashes with that very simple code in Mix_Close_Audio():
```
#include <stdio.h>
#include <SDL.h>
#include <SDL_mixer.h>

int main ( int argc, char *argv[] )
{
	int result;
  
	SDL_Init(SDL_INIT_AUDIO);
	result = Mix_OpenAudio(22050, AUDIO_S16LSB, 2, 1024);
	if(result>=0) {
		puts("check\n");
		Mix_CloseAudio();
		puts("ok\n");
	}
	SDL_Quit();
	return(0);
}

```
I have recompiled a new libSDL_mixer.a with the latest gcc, and that sample program doesn't seem to crash. I now need to check (the much bigger) DevilutionX... stay tuned.

**Comment by bebbo** [comment]
do you plan to publish the recipe how to build DevilutionX with gcc?

**Comment by Samuel-DEVULDER** [comment]
I'm not sure to undertand the question. To build I just cd in the "build" folder and "make". The trickiest part is setting up the dev env for which I had to get help from [Arczi](https://github.com/arczi84/devilutionX). There is a "prep.sh" script in "Packaging/amiga" which download & compile missing libs, then I go into the "build" folder and run "CMake/amiga/Configure.readme" which build the Makefiles. Then I can compile with "make" (or "make VERBOSE=1" to see what's going on) from withing that "build" folder. In case of doubt I can "rm -rf" that folder, and redo the whole configure process from scratch. But for this version, it seem I also had to rebuild libSDL_mixer.a. I don't see why it crash because there is no amiga-specific stuff in there... so it is still a bit mysterious to me. Let's hope I don't have to rebuild all the extra libs (there are quite a few, and the prep.h script doen't work "out of the box" on my setup.)
